File: Daily Challenge/March 2025/Minimum Time to Repair Cars.cpp
class Solution {
public:
  long long repairCars(vector<int>& ranks, int cars) {
    long long left = 1, right = 1e16;

    while (left <= right) {
      long long mid = (left + right) / 2;

      if (canRepair(ranks, cars, mid)) {
        right = mid - 1;
      } else {
        left = mid + 1;
      }
    }

    return left;
  }

private:
  bool canRepair(vector<int>& ranks, int cars, long long time) {
    long long carCount = 0;

    for (int i = 0; i < ranks.size(); ++i) {
      long long repair = time / ranks[i];
      repair = sqrt(repair);
      carCount += repair;
    }

    return carCount >= cars;
  }
};

File: Daily Challenge/March 2025/Minimum Time to Repair Cars.cs
public class Solution {
  public long RepairCars(int[] ranks, int cars) {
    long left = 1, right = (long) 1e16;

    while (left <= right) {
      long mid = (left + right) / 2;

      if (CanRepair(ranks, cars, mid)) {
        right = mid - 1;
      } else {
        left = mid + 1;
      }
    }

    return left;
  }

  private bool CanRepair(int[] ranks, int cars, long time) {

    long carsRepaired = 0;

    for (int i = 0; i < ranks.Length; ++i) {
      long repairs = time / ranks[i];
      repairs = (long) Math.Sqrt(repairs);
      carsRepaired += repairs;
    }

    return carsRepaired >= cars;
  }
}